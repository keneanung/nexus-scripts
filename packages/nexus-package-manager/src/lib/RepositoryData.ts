export type RepositoryData = PackageEntry[];
export interface PackageEntry {
  name: string;
  packageName: string;
  description: string;
  url: string;
  dependencies: string[];
  website?: string;
  version: string;
}
