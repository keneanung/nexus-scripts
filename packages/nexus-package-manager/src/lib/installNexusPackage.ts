export const installNexusPackage = (pkgJson: any) => {
  const newPkg = globalThis.nexusclient.packages().create(pkgJson.name, pkgJson.description);
  const reflexes = globalThis.nexusclient.reflexes();
  newPkg.apply(pkgJson, reflexes);
  reflexes.run_function('onInstall', undefined, pkgJson.name);
};
