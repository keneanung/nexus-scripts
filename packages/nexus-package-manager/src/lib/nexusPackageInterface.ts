export const installNexusPackage = (pkgJson: any) => {
  const newPkg = globalThis.nexusclient.packages().create(pkgJson.name, pkgJson.description);
  const reflexes = globalThis.nexusclient.reflexes();
  newPkg.apply(pkgJson, reflexes);
  reflexes.run_function('onInstall', undefined, pkgJson.name);
};

export const isNexusPackageInstalled = (packageName: string) => {
  return globalThis.nexusclient.packages().exists(packageName, true) !== 0;
};

export const uninstallNexusPackage = (packageName: string) => {
  globalThis.nexusclient.reflexes().run_function('onUninstall', undefined, packageName);
  globalThis.nexusclient.packages().remove(packageName);
};

export const getPackageList = (): nexusclient.ReflexPackage[] => {
  return nexusclient.reflexes().get_package_list();
};

export const setPackageOrder = (packageList: nexusclient.ReflexPackage[]) => {
  for (let i = 0; i < packageList.length; i++) {
    const packageList = getPackageList();
    const oldIndex = packageList.findIndex((p: nexusclient.ReflexPackage) => p.name === packageList[i].name);
    nexusclient.packages().move(oldIndex, i);
  }
};
