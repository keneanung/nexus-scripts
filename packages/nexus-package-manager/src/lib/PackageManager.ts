import { installNexusPackage } from "./installNexusPackage";
import { RepositoryData } from "./RepositoryData";

export class PackageManager{

  private repositoryData: RepositoryData = [];

  public updateAsync = async () => {
    const response = await fetch('https://keneanung.github.io/nexus-package-repository/repository.json');
    this.repositoryData = await response.json();
  }

  public getRepositoryData = () => this.repositoryData;
  public installAsync = async (packageName: string) => {
    const pkgEntry = this.repositoryData.find(entry => entry.packageName === packageName);
    if(!pkgEntry){
      throw new Error("");

    }
    const response = await fetch(pkgEntry.url);
    const pkgJson = await response.json();

    installNexusPackage(pkgJson);
  }
}
