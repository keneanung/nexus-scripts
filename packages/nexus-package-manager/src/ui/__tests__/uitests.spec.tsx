import { PackageManager } from '../../lib/PackageManager';
import { PackageManagerUi } from '../index';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { isNexusPackageInstalled } from '../../lib/nexusPackageInterface';
jest.mock('../../lib/nexusPackageInterface');
const isNexusPackageInstalledMock = jest.mocked(isNexusPackageInstalled);

const server = setupServer(
  rest.get('https://keneanung.github.io/nexus-package-repository/repository.json', (req, res, ctx) => {
    return res(
      ctx.json([
        {
          dependencies: [],
          description: 'foo',
          name: 'bar',
          packageName: 'barPackage',
          url: 'https://mykg.com/bar.json',
        },
      ]),
    );
  }),
  rest.get('https://mykg.com/bar.json', (req, res, ctx) => {
    return res(ctx.json([]));
  }),
);

// Enable request interception.
beforeAll(() =>
  server.listen({
    onUnhandledRequest: 'error',
  }),
);

// Reset handlers so that each test could alter them
// without affecting other, unrelated tests.
afterEach(() => server.resetHandlers());

// Don't forget to clean up afterwards.
afterAll(() => server.close());

test('Should have an empty table without package data', () => {
  const packageManager = new PackageManager();

  const { container } = render(<PackageManagerUi packageManager={packageManager} />);

  expect(container).toMatchSnapshot();
});

test('Should have a table with single package data to install if one package was provided and not installed', async () => {
  isNexusPackageInstalledMock.mockReturnValue(false);
  const packageManager = new PackageManager();
  const { container } = render(<PackageManagerUi packageManager={packageManager} />);

  fireEvent.click(screen.getByText('Update package listing'));
  await waitFor(() => expect(screen.queryAllByRole('row')).toHaveLength(2));

  expect(container).toMatchSnapshot();
});

test('Should have a table with single package data to update if one package was provided and installed', async () => {
  isNexusPackageInstalledMock.mockReturnValue(true);
  const packageManager = new PackageManager();
  const { container } = render(<PackageManagerUi packageManager={packageManager} />);

  fireEvent.click(screen.getByText('Update package listing'));
  await waitFor(() => expect(screen.queryAllByRole('row')).toHaveLength(2));

  expect(container).toMatchSnapshot();
});

test('Should have a table with single package data to update if one package was provided and installed after clicking the button', async () => {
  isNexusPackageInstalledMock.mockReturnValue(false);
  const packageManager = new PackageManager();
  const { container } = render(<PackageManagerUi packageManager={packageManager} />);

  fireEvent.click(screen.getByText('Update package listing'));
  await waitFor(() => expect(screen.queryAllByRole('row')).toHaveLength(2));
  fireEvent.click(screen.getByText('Install'));
  await waitFor(() => expect(screen.getByText('Uninstall')).toBeDefined());

  expect(container).toMatchSnapshot();
});

test('Should have a table with single package data to install if one package was provided and uninstalled after clicking the button', async () => {
  isNexusPackageInstalledMock.mockReturnValue(true);
  const packageManager = new PackageManager();
  const { container } = render(<PackageManagerUi packageManager={packageManager} />);

  fireEvent.click(screen.getByText('Update package listing'));
  await waitFor(() => expect(screen.queryAllByRole('row')).toHaveLength(2));
  fireEvent.click(screen.getByText('Uninstall'));
  await waitFor(() => expect(screen.getByText('Install')).toBeDefined());

  expect(container).toMatchSnapshot();
});
