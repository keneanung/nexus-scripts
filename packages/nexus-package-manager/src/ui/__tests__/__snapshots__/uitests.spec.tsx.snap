// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should have a table with single package data to install if one package was provided and not installed 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
  >
    Update package listing
  </button>
  <table>
    <thead>
      <tr>
        <td>
          Name
        </td>
        <td>
          Package Name
        </td>
        <td>
          Description
        </td>
        <td>
          Actions
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          bar
        </td>
        <td>
          barPackage
        </td>
        <td>
          foo
        </td>
        <td>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          >
            Install
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Should have a table with single package data to install if one package was provided and uninstalled after clicking the button 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
  >
    Update package listing
  </button>
  <table>
    <thead>
      <tr>
        <td>
          Name
        </td>
        <td>
          Package Name
        </td>
        <td>
          Description
        </td>
        <td>
          Actions
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          bar
        </td>
        <td>
          barPackage
        </td>
        <td>
          foo
        </td>
        <td>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          >
            Install
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Should have a table with single package data to update if one package was provided and installed 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
  >
    Update package listing
  </button>
  <table>
    <thead>
      <tr>
        <td>
          Name
        </td>
        <td>
          Package Name
        </td>
        <td>
          Description
        </td>
        <td>
          Actions
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          bar
        </td>
        <td>
          barPackage
        </td>
        <td>
          foo
        </td>
        <td>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          >
            Update
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          >
            Uninstall
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Should have a table with single package data to update if one package was provided and installed after clicking the button 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
  >
    Update package listing
  </button>
  <table>
    <thead>
      <tr>
        <td>
          Name
        </td>
        <td>
          Package Name
        </td>
        <td>
          Description
        </td>
        <td>
          Actions
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          bar
        </td>
        <td>
          barPackage
        </td>
        <td>
          foo
        </td>
        <td>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          >
            Update
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          >
            Uninstall
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Should have an empty table without package data 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
  >
    Update package listing
  </button>
  <table>
    <thead>
      <tr>
        <td>
          Name
        </td>
        <td>
          Package Name
        </td>
        <td>
          Description
        </td>
        <td>
          Actions
        </td>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
